Revert "Revert "[execute_process] emulate_tty configurable and defaults to true""

wjwwood opened the pull request:
This is a refactor of https://github.com/ros2/launch/pull/265 so that it doesn't break our other tests.

Detecting it in the first place was my fault because I narrowed our CI testing too much and so didn't exercise the failing tests.

Reverts ros2/launch#276

dirk-thomas commented:
If want a fast CI response for this change you can try the new CI overlay job. It uses all unchanged packages from last nights build and only builds the specified changes: see http://build.ros2.org/view/Eci/job/Eci__overlay_ubuntu_bionic_amd64/21/parameters/ which finished in less than 15 min.

wjwwood commented:
Ok, I found the issue, it was the ANSI escape sequences (printed by the console logger to reset terminal colors) were being captured when emulate tty was enabled, which broke output comparison. I updated launch testing to strip these by default (with an option to not strip them in case someone needed that feature later).

@pbaughman can you look at 12c0563? I moved this around because I noticed it and thought it might be a race where the event handlers could miss some events from processes included in the launch description being tested. I don't think it was the issue, but while debugging I considered it might be an issue. I can revert it if needed, but I think the new code is more correct.

wjwwood commented:
Here's full CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7626)](http://ci.ros2.org/job/ci_linux/7626/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3742)](http://ci.ros2.org/job/ci_linux-aarch64/3742/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6235)](http://ci.ros2.org/job/ci_osx/6235/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=7473)](http://ci.ros2.org/job/ci_windows/7473/)

dirk-thomas commented:
A CI build on build.ros2.org shows the same test failures as previously: http://build.ros2.org/view/Eci/job/Eci__overlay_ubuntu_bionic_amd64/25/testReport/

wjwwood commented:
Configuration looks ok, not sure. It definitely fixed it in my local vm and I pushed everything. The job used the right commit of ros2/launch (afaict).

The failing test names and captured output are kind of not very useful though.

wjwwood commented:
Is it possible that the overlaid `demo_nodes_cpp` is using the `launch` from the underlay?

dirk-thomas commented:
> Is it possible that the overlaid `demo_nodes_cpp` is using the `launch` from the underlay?

It shouldn't. But I wouldn't go as far as to say that it is impossible :wink:

wjwwood commented:
It's possible this fix isn't perfect, let's see what ci.ros2.org says, and I'll investigate with CI jobs containing print statements in launch if needed.

wjwwood commented:
Looks like it is failing on ci.ros2.org too (just reading the output so far), I guess I'll have to have another look. It might be the `\n` versus `\r\n`, but I don't know why my VM works though.

wjwwood commented:
Ok, all the tests in `demo_nodes_cpp` were fixed, but none of those used the regex feature :|. So now I changed how newline replacement works because when emulating a tty, `\r\n` is used as the line ending rather than just `\n` or `os.linesep` as it is in files. I updated the logic to handle this. We do the same in `osrf_pycommon`: https://github.com/osrf/osrf_pycommon/blob/8ba78744d9df7de70d824b4d76c11afe9e81ccd1/osrf_pycommon/process_utils/execute_process_pty.py#L80-L81

But it's still failing on the build.ros2.org ci: http://build.ros2.org/view/Eci/job/Eci__overlay_ubuntu_bionic_amd64/26/console

I'm not sure what's happening just yet.

wjwwood commented:
Running a smaller set of CI on ci.ros2.org, because I think it's possible that nothing is fixed on build.ros.org (`demo_nodes_cpp` nor `intra_process_demo`) but that before `demo_nodes_cpp` was fixed on ci.ros2.org and only `intra_process_demo` and other regex based tests were still failing:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7629)](http://ci.ros2.org/job/ci_linux/7629/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3744)](http://ci.ros2.org/job/ci_linux-aarch64/3744/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6237)](http://ci.ros2.org/job/ci_osx/6237/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=7476)](http://ci.ros2.org/job/ci_windows/7476/)

wjwwood commented:
Yeah, looks like the build.ros2.org CI job is not catching the fix some how, the `demo_nodes_cpp` issues are fixed in https://ci.ros2.org/job/ci_linux/7626/testReport/ (from the first set of CI I started on ci.ros2.org).

pbaughman commented:
> @pbaughman can you look at 12c0563? I moved this around because I noticed it and thought it might be a race where the event handlers could miss some events from processes included in the launch description being tested. I don't think it was the issue, but while debugging I considered it might be an issue. I can revert it if needed, but I think the new code is more correct.

@wjwwood I never considered that the order of event handlers and execute processes could matter but yeah - you're right.  In a perfect world, I would hope it wouldn't matter but in the real world, yours is almost certainly more correct.

wjwwood commented:
I need to still look at this failure (last one):

https://ci.ros2.org/job/ci_linux/7629/testReport/launch.test.launch.actions/test_include_launch_description/test_include_launch_description_launch_arguments/

But it's failing on nightly too:

https://ci.ros2.org/view/nightly/job/nightly_linux_release/1245/testReport/launch.test.launch.actions/test_include_launch_description/test_include_launch_description_launch_arguments/

So, I'd say this is up for review.

hidmic commented on launch/launch/actions/execute_process.py:
```diff
@@ -577,6 +583,16 @@ def __expand_substitutions(self, context):
+                    type(emulate_tty)
+                ))
+        except KeyError:
+            emulate_tty = self.__emulate_tty
```
I'm usually OK with Python's style of simply letting things throw but in this case I think that checking for `'emulate_tty' in context.launch_configurations` instead would aid readability.

wjwwood replied:
Have a look at 1b84e42.

hidmic commented on launch_testing/launch_testing/asserts/assert_output.py:
```diff
@@ -16,6 +16,19 @@
+    lines = lines.replace(os.linesep, '\n')
+    # This happens (even on Linux and macOS) when capturing I/O from an
+    # emulated tty.
+    lines = lines.replace('\r\n', '\n')
```
@wjwwood hmm, funky... Do we know why?

wjwwood replied:
Yeah, it uses `\r` to move the carriage return back to the beginning of the line. Most print functions emulate this behavior for you when you're connected to a tty and print a lone `\n`.

It's a quirk of these kinds of things, which I've mostly learned from experience. There's scarcely any documentation about these behaviors across operating systems and different configurations. So yeah, funky.

hidmic commented on launch/test/launch/actions/test_emulate_tty.py:
```diff
@@ -0,0 +1,70 @@
+    # override the default via LaunchConfiguration
+    ('false', 0)
+])
+def test_emulate_tty(test_input, expected):
```
@wjwwood it'd be nice to test that passing `emulate_tty` upon `ExecuteProcess` construction works too.

wjwwood replied:
Check 9b413c1.

hidmic commented on launch/launch/actions/execute_process.py:
```diff
@@ -577,6 +583,16 @@ def __expand_substitutions(self, context):
+        try:
+            emulate_tty = yaml.safe_load(
+                context.launch_configurations['emulate_tty']
+            )
```
@wjwwood meta: I think `yaml` parsing here is fine for now, but I also think we need to get this straight in `launch`. We're already applying YAML parsing under the hood for parameters, maybe we should standardize it (or not)?

wjwwood replied:
In the original pull request I was going to mention this, but I didn't just to keep things moving. We have a function for this kind of parsing which we use in the condition evaluations:

https://github.com/ros2/launch/blob/master/launch/launch/conditions/evaluate_condition_expression_impl.py

It's a subset of YAML. I think we can/should stabilize on something. I don't mind using yaml, though we would need to depend on pyyaml. Which brings up another issue which is that launch doesn't declare a dependency on yaml after this pr. So that's not right. I think this is why I made the evaluation function in the first place. I think I will reuse that and drop the use of YAML for now. Allowing `true`(any cap)/`1`/`false`/`0` is documented in other parts of the code too.

wjwwood replied:
See 1b84e42

wjwwood commented:
CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7635)](http://ci.ros2.org/job/ci_linux/7635/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3749)](http://ci.ros2.org/job/ci_linux-aarch64/3749/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6242)](http://ci.ros2.org/job/ci_osx/6242/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=7482)](http://ci.ros2.org/job/ci_windows/7482/)

wjwwood commented:
@stonier I refactored two parts of the original pull request, just FYI in case you care to have a look.

wjwwood commented:
9b413c1 also fixes the test failure that is also on master for the nightly's.

stonier commented:
Looks good to me - nicer way of checking the boolean.

wjwwood commented:
I'm looking into https://ci.ros2.org/job/ci_linux/7635/testReport/(root)/projectroot/test_logging_output_format/ now, which also seems to be related to this change.

wjwwood commented:
So this last issue has become a bit crazy. I'm still working on it, but I'm going to change the default of `emulate_tty` to be `False` (current behavior), and make a new pull request to change it to be `True` by default along with any fixes needed to make the last kind of failing test pass.

wjwwood commented:
CI:

* Linux [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux&build=7654)](http://ci.ros2.org/job/ci_linux/7654/)
* Linux-aarch64 [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_linux-aarch64&build=3765)](http://ci.ros2.org/job/ci_linux-aarch64/3765/)
* macOS [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_osx&build=6261)](http://ci.ros2.org/job/ci_osx/6261/)
* Windows [![Build Status](http://ci.ros2.org/buildStatus/icon?job=ci_windows&build=7498)](http://ci.ros2.org/job/ci_windows/7498/)

wjwwood commented:
@stonier FYI I had to disable emulate_tty by default right now, so that I can get some fixes in and at least have the feature available. I'll keep poking at the issue and try to get it so that we can re-enable it by default.

wjwwood commented:
CI test failures are from new overnight linter warnings.

stonier commented:
> @stonier FYI I had to disable emulate_tty by default right now, so that I can get some fixes in and at least have the feature available. I'll keep poking at the issue and try to get it so that we can re-enable it by default.

Understood, thanks.
